common/src/main/scala/net/psforever/objects/serverobject/doors/Door.scala

```diff
@@ -92,4 +122,25 @@ object Door {
     obj.Actor = context.actorOf(Props(classOf[DoorControl], obj), s"${GlobalDefinitions.door.Name}_$id")
     obj
   }
+
+  import net.psforever.types.Vector3
+  /**
+    * Instantiate and configure a `Door` object that has knowledge of both its position and outwards-facing direction.
+    * The assumption is that this door will be paired with an IFF Lock, thus, has conditions for opening.
+    * @param pos the position of the door
+    * @param outwards_direction a vector in the direction of the door's outside
+    * @param id the unique id that will be assigned to this entity
+    * @param context a context to allow the object to properly set up `ActorSystem` functionality
+    * @return the `Door` object
+    */
+  def Constructor(pos : Vector3, outwards_direction : Vector3)(id : Int, context : ActorContext) : Door = {
```

Comment:
Why are the first two parameters separated from the last two?

Reply:
This is a form of currying.  A typical function can be defined as `(A, B) => C`, where `A` and `B` are the parameter types to a function and `C` is the returned type.  When you do `func(A, B)`, it executes and produces a `C`.  Separating the parameters like this defines an `A => B => C` function - an `(A) => ((B) => C)` function, if you will - that can accept an `A` parameter and outputs a `(B) => C` function.  Its execution has been deferred.

What I am essentially doing is providing the function with some of its parameters in a place where it can reliably expect those parameters and then putting the function somewhere until I need it later.  What I am waiting on is an appropriate `ActorContext` so the actor of this object is created under a particular actor hierarchy.  I have not yet built the basic object because I can make a new object each time I call this function which will be useful to avoid some config redundancy (zones using the same map).

Reply:
Ah, I was wondering if that was a curry. At my job I have been working with React.js related libraries that use currying such as Redux and Recompose (if you are familiar with all the Javascript craziness), but I was not familiar with this Scala syntactic sugar.

So this lets you go ahead and create a `Door` before you have the appropriate `ActorContext`, but more importantly, it let's you easily create multiple instances of a `Door` with the given `pos` and `outwards_direction` with different `ActorContext`'s?

Reply:
Yes.  Consider the VR training in the game - the same map layout for different faction-separated zones.  In our case, we'd use the same map to set up VR driving zones `tzdrtr`, `tzdrnc`, and `tzdrvs`.

I could easily hardcode everything into the `Zone` objects and make the `ZoneMap` class and the currying superfluous.  There's only nine normal zones - three layouts - that would feature the duplication.  From a purist perspective, that would be the most definite of forms because it would be hard coded.  I'm holding out for the enthusiasm to write/include a proper no-compile-file configuration method, such as something with XML, however.  That's why the main point for me is the ability to separate use of more primitive configuration data from use of the Actor system wiring.

I feel I have to prepare for the ambitious people who want to expand on the game after we finally get to vanilla PlanetSide Classic.